libdnf5/conf/option_enum.cpp

```diff
@@ -22,169 +22,89 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include "libdnf5/utils/bgettext/bgettext-mark-domain.h"
 
 #include <algorithm>
-#include <sstream>
 
 namespace libdnf5 {
 
-template <typename T>
-bool from_string(T & out, const std::string & in, std::ios_base & (*manipulator)(std::ios_base &)) {
-    std::istringstream iss(in);
-    return !(iss >> manipulator >> out).fail();
-}
-
-template <typename T>
-OptionEnum<T>::OptionEnum(ValueType default_value, const std::vector<ValueType> & enum_vals)
+class OptionEnum::Impl {
+public:
+    Impl(std::string && default_value, std::vector<std::string> && enum_vals)
+        : enum_vals(std::move(enum_vals)),
+          default_value(default_value),
+          value(std::move(default_value)) {}
+
+    Impl(std::string && default_value, std::vector<std::string> && enum_vals, FromStringFunc && from_string_func)
+        : from_string_user(std::move(from_string_func)),
+          enum_vals(std::move(enum_vals)),
+          default_value(default_value),
+          value(std::move(default_value)) {}
+
+private:
+    friend OptionEnum;
+
+    FromStringFunc from_string_user;
+    std::vector<std::string> enum_vals;
+    std::string default_value;
+    std::string value;
+};
+
+OptionEnum::OptionEnum(std::string default_value, std::vector<std::string> enum_vals)
     : Option(Priority::DEFAULT),
-      enum_vals(enum_vals),
-      default_value(default_value),
-      value(default_value) {
+      p_impl(new Impl(std::move(default_value), std::move(enum_vals))) {
     test(default_value);
```

Comment:
Thanks for the changes.
Please replace `test(default_value);` by `test(p_impl->value);` as it is in my code. `default_value` has been moved to `p_impl->value` and no longer heva a valid value.

Reply:
Thanks, I have updated it.